src/persist-client/src/lib.rs

```diff
@@ -0,0 +1,315 @@
+    // can definitely revisit this later if it ends up being a bottleneck.
+    //
+    // WIP: This is an opportunity to push down Projection information to save
+    // bandwidth. My thinking so far is that Map and Filter should stay
+    // STORAGE-level concerns.
+    //
+    // WIP: My first attempt here was to return `impl Stream` but that didn't
+    // compile with `todo!`. I don't yet have an opinion on which is better.
+    pub async fn tail(as_of: Antichain<T>) -> Result<Tail<K, V, T, D>, Error> {
+        todo!("{:?}", as_of);
+    }
+
+    // NB the len of the returned vec is num_shards
+    //
+    // WIP This is an opportunity to push down Projection information to save
+    // bandwidth. My thinking so far is that Map and Filter should stay
+    // STORAGE-level concerns.
+    pub async fn snapshot(
+        &self,
+        as_of: Antichain<T>,
+        num_shards: NonZeroUsize,
+    ) -> Result<Vec<SnapshotShard>, Error> {
+        todo!("{:?}{:?}", as_of, num_shards);
+    }
+
+    // WIP: Should this live on ReadHandle or somewhere else? E.g. I think it
+    // could also live on Client.
+    //
+    // WIP: Figure out the error story here. Retries should move inside this, but
+    // what happens if we get data that we can't decode?
+    //
+    // WIP: My first attempt here was to return `impl Stream` but that didn't
+    // compile with `todo!`. I don't yet have an opinion on which is better.
+    pub async fn snapshot_shard(
+        &self,
+        shard: SnapshotShard,
+    ) -> Result<SnapshotIter<K, V, T, D>, Error> {
+        todo!("{:?}", shard);
+    }
+
+    pub async fn clone(&self) -> Result<Self, Error> {
+        todo!();
+    }
+}
+
+impl<K, V, T, D> ReadHandle<K, V, T, D> {
+    pub async fn deregister(&mut self) -> Result<(), Error> {
+        todo!()
+    }
+}
+
+impl<K, V, T, D> Drop for ReadHandle<K, V, T, D> {
+    fn drop(&mut self) {
+        // WIP: Thread a tokio runtime down instead?
+        futures_executor::block_on(async {
+            if let Err(err) = self.deregister().await {
+                warn!("failed to deregister ReadHandle on drop: {}", err)
+            }
+        })
+    }
+}
+
+pub struct WriteHandle<K, V, T, D> {
+    _phantom: PhantomData<(K, V, T, D)>,
+}
+
+impl<K, V, T, D> WriteHandle<K, V, T, D>
+where
+    K: Codec,
+    V: Codec,
+    T: DurableTimestamp,
+    D: DurableDiff,
+{
+    // This handle's upper frontier, not the global collection-level one.
+    pub fn upper(&self) -> &Antichain<T> {
+        todo!()
+    }
+
+    // NB new_upper of the empty antichain "finishes" this collection, promising
+    // that no more data is ever incoming
+    //
+    // TODO: Bound the write and space amplification as a function of how much
+    // data has even been written to this collection as well as how much of it
+    // is "live" (doesn't consolidate out at the collection frontier).
+    pub async fn write_batch<I: IntoIterator<Item = ((K, V), T, D)>>(
```

Comment:
WIP: Document latency expectations. Don't focus on this too much but an initial guess is something like "p50 < 100ms but p95/p99 > 1s". However, throughput here should be "nearly saturate your network" good. Now that I bring this up, I wonder if it's worth bringing over the current-persist separation of "received your write, ready for your next write" vs "your write is durable" (concretely: getting a PFuture from StreamWriteHandle::write is the first, the PFuture resolving a value is the second). We'll certainly need something like that to maximize throughput.